services/brig/src/Brig/API/OAuth.hs

```diff
@@ -0,0 +1,397 @@
+    mkRefreshToken = do
+      rid :: OAuthRefreshTokenId <- randomId
+      sub <- maybe (throwStd $ errorToWai @'OAuthJwtError) pure $ idToText rid ^? stringOrUri
+      let claims = emptyClaimsSet & claimSub ?~ sub
+      (rid,) . OAuthToken <$> signRefreshToken claims
+
+    mkAccessToken :: Member Now r => (Handler r) OAuthAccessToken
+    mkAccessToken = do
+      domain <- Opt.setFederationDomain <$> view settings
+      exp <- fromIntegral . Opt.setOAuthAccessTokenExpirationTimeSecs <$> view settings
+      claims <- mkAccessTokenClaims uid domain scope exp
+      OAuthToken <$> signAccessToken claims
+
+    mkAccessTokenClaims :: Member Now r => UserId -> Domain -> OAuthScopes -> NominalDiffTime -> (Handler r) OAuthClaimsSet
+    mkAccessTokenClaims u domain scopes ttl = do
+      iat <- lift (liftSem Now.get)
+      uri <- maybe (throwStd $ errorToWai @'OAuthJwtError) pure $ domainText domain ^? stringOrUri
+      sub <- maybe (throwStd $ errorToWai @'OAuthJwtError) pure $ idToText u ^? stringOrUri
+      let exp = addUTCTime ttl iat
+      let claimSet =
+            emptyClaimsSet
+              & claimIss ?~ uri
+              & claimAud ?~ Audience [uri]
+              & claimIat ?~ NumericDate iat
+              & claimSub ?~ sub
+              & claimExp ?~ NumericDate exp
+      pure $ OAuthClaimsSet claimSet scopes
+
+    signAccessToken :: OAuthClaimsSet -> (Handler r) SignedJWT
+    signAccessToken claims = do
+      jwtOrError <- liftIO $ doSignClaims
+      either (const $ throwStd $ errorToWai @'OAuthJwtError) pure jwtOrError
+      where
+        doSignClaims :: IO (Either JWTError SignedJWT)
+        doSignClaims = runJOSE $ do
+          algo <- bestJWSAlg key
+          signJWT key (newJWSHeader ((), algo)) claims
+
+    signRefreshToken :: ClaimsSet -> (Handler r) SignedJWT
+    signRefreshToken claims = do
+      jwtOrError <- liftIO $ doSignClaims
+      either (const $ throwStd $ errorToWai @'OAuthJwtError) pure jwtOrError
+      where
+        doSignClaims :: IO (Either JWTError SignedJWT)
+        doSignClaims = runJOSE $ do
+          algo <- bestJWSAlg key
+          signClaims key (newJWSHeader ((), algo)) claims
+
+verifyClientSecret :: OAuthClientPlainTextSecret -> OAuthClientId -> (Handler r) Bool
+verifyClientSecret secret cid = do
+  let plainTextPw = PlainTextPassword $ toText $ unOAuthClientPlainTextSecret secret
+  lift $
+    wrapClient $
+      lookupOAuthClientSecret cid <&> \case
+        Nothing -> False
+        Just pw -> verifyPassword plainTextPw pw
+
+rand32Bytes :: MonadIO m => m AsciiBase16
+rand32Bytes = liftIO . fmap encodeBase16 $ randBytes 32
+
+--------------------------------------------------------------------------------
+
+revokeRefreshToken :: Member Jwk r => OAuthRevokeRefreshTokenRequest -> (Handler r) ()
+revokeRefreshToken req = do
+  key <- signingKey
+  info <- verifyTokenAndGetInfo key (ortrRefreshToken req)
```

Comment:
`verifyTokenAndGetInfo` removes the authorization token.  in what context does the oauth client still have the authorization token, it's not deleted on wire-server yet, and the oauth client has a refresh token that it wants to revoke?